File: android/app/src/main/java/com/hatem/musicmute/processing/ProcessingUsageRepository.kt
package com.hatem.musicmute.processing

import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

@Serializable
data class ProcessingUsagePeriod(
    val key: String,
    val start: String,
    val end: String,
    val nextResetAt: String,
)

@Serializable
data class ProcessingUsageAmounts(
    val limitSeconds: Double,
    val usedSeconds: Double,
    val reservedSeconds: Double,
    val releasedSeconds: Double,
    val remainingSeconds: Double,
)

@Serializable
data class ProcessingUploadUsage(
    val dailyGrantLimit: Int,
    val dailyGrants: Int,
    val dailyRemainingGrants: Int,
    val dailyResetAt: String,
    val monthlyGrantLimit: Int,
    val monthlyGrants: Int,
    val monthlyRemainingGrants: Int,
    val monthlyByteLimit: Long,
    val confirmedBytes: Long,
    val monthlyRemainingBytes: Long,
    val monthlyResetAt: String,
)

@Serializable
data class ProcessingStorageUsage(
    val limitBytes: Long,
    val retainedBytes: Long,
    val remainingBytes: Long,
)

@Serializable
data class ProcessingDownloadUsage(
    val monthlyGrantLimit: Int,
    val monthlyGrants: Int,
    val monthlyRemainingGrants: Int,
    val monthlyByteLimit: Long,
    val estimatedBytes: Long,
    val monthlyRemainingBytes: Long,
    val monthlyResetAt: String,
)

@Serializable
data class ProcessingEffectiveLimits(
    val maxDurationSeconds: Int,
    val maxPreparedAudioBytes: Long,
    val maxClientInputAttempts: Int,
    val signedUrlTtlSeconds: Int,
)

@Serializable
data class ProcessingAvailability(val status: String, val reason: String? = null)

@Serializable
data class ProcessingUsage(
    val schemaVersion: Int,
    val plan: String,
    val policyRevision: Int,
    val overrideRevision: Int? = null,
    val effectivePolicySource: String,
    val overrideExpiresAt: String? = null,
    val period: ProcessingUsagePeriod,
    val processing: ProcessingUsageAmounts,
    val uploads: ProcessingUploadUsage,
    val storage: ProcessingStorageUsage,
    val effectiveLimits: ProcessingEffectiveLimits,
    val downloads: ProcessingDownloadUsage,
    val usageRevision: Int,
    val waitingJobs: Int,
    val maxWaitingJobs: Int,
    val processingJobs: Int,
    val maxProcessingJobs: Int,
    val availability: ProcessingAvailability,
    val checkedAt: String,
) {
    fun validate() {
        val amounts = listOf(
            processing.limitSeconds,
            processing.usedSeconds,
            processing.reservedSeconds,
            processing.releasedSeconds,
            processing.remainingSeconds,
        )
        if (schemaVersion != 2 || plan != "standard" || policyRevision < 0 ||
            overrideRevision?.let { it < 1 } == true || usageRevision < 0 ||
            effectivePolicySource !in setOf("global", "account_override") ||
            amounts.any { !it.isFinite() || it < 0 } ||
            processing.remainingSeconds > processing.limitSeconds ||
            listOf(
                uploads.dailyGrantLimit, uploads.monthlyGrantLimit,
                downloads.monthlyGrantLimit, effectiveLimits.maxDurationSeconds,
                effectiveLimits.maxClientInputAttempts, effectiveLimits.signedUrlTtlSeconds,
            ).any { it < 1 } ||
            listOf(
                uploads.dailyGrants, uploads.dailyRemainingGrants, uploads.monthlyGrants,
                uploads.monthlyRemainingGrants, downloads.monthlyGrants,
                downloads.monthlyRemainingGrants,
            ).any { it < 0 } ||
            listOf(
                uploads.monthlyByteLimit, uploads.confirmedBytes, uploads.monthlyRemainingBytes,
                storage.limitBytes, storage.retainedBytes, storage.remainingBytes,
                downloads.monthlyByteLimit, downloads.estimatedBytes,
                downloads.monthlyRemainingBytes, effectiveLimits.maxPreparedAudioBytes,
            ).any { it < 0 } ||
            uploads.dailyRemainingGrants > uploads.dailyGrantLimit ||
            uploads.monthlyRemainingGrants > uploads.monthlyGrantLimit ||
            uploads.monthlyRemainingBytes > uploads.monthlyByteLimit ||
            storage.remainingBytes > storage.limitBytes ||
            downloads.monthlyRemainingGrants > downloads.monthlyGrantLimit ||
            downloads.monthlyRemainingBytes > downloads.monthlyByteLimit ||
            effectiveLimits.signedUrlTtlSeconds > 600 ||
            waitingJobs < 0 || maxWaitingJobs < 1 ||
            processingJobs < 0 || maxProcessingJobs < 1 ||
            availability.status !in setOf("available", "blocked") ||
            availability.reason !in setOf(null, "paused", "monthly_limit_reached", "waiting_job_limit", "storage_limit_reached") ||
            (availability.status == "available") != (availability.reason == null)) {
            throw JobsFailure(JobsProblem.SERVICE_UNAVAILABLE)
        }
        try {
            val start = Instant.parse(period.start)
            val end = Instant.parse(period.end)
            require(start < end && Instant.parse(period.nextResetAt) == end)
            Instant.parse(uploads.dailyResetAt)
            require(Instant.parse(uploads.monthlyResetAt) == end)
            require(Instant.parse(downloads.monthlyResetAt) == end)
            Instant.parse(checkedAt)
            overrideExpiresAt?.let(Instant::parse)
        } catch (_: Exception) { throw JobsFailure(JobsProblem.SERVICE_UNAVAILABLE) }
    }
    fun requireAvailable(checkAvailability: Boolean = true) {
        if (waitingJobs >= maxWaitingJobs) throw JobsFailure(JobsProblem.PROCESSING_LIMIT_REACHED)
        if (processing.remainingSeconds <= 0) throw JobsFailure(JobsProblem.PROCESSING_ALLOWANCE_EXHAUSTED)
        if (!checkAvailability) return
        when (availability.reason) {
            "monthly_limit_reached" -> throw JobsFailure(JobsProblem.PROCESSING_ALLOWANCE_EXHAUSTED)
            "waiting_job_limit" -> throw JobsFailure(JobsProblem.PROCESSING_LIMIT_REACHED)
            "paused" -> throw JobsFailure(JobsProblem.PROCESSING_CAPACITY_UNAVAILABLE)
        }
    }
}

/** Owner-fenced, nonbinding preflight. Never reserves capacity before local preparation. */
class ProcessingUsageRepository(private val api: JobsApi, private val session: () -> ProcessingSession?) {
    private val mutable = MutableStateFlow<ProcessingUsage?>(null)
    val usage = mutable.asStateFlow()
    suspend fun refresh(): ProcessingUsage? {
        val owner = session() ?: run { mutable.value = null; return null }
        mutable.value = null
        val result = try { api.processingUsage() } catch (error: CancellationException) { throw error }
            catch (error: JobsFailure) {
                // Legacy servers and disconnected clients keep their safe local flow;
                // actual admission still authoritatively enforces access and capacity.
                if (error.problem in setOf(JobsProblem.JOB_NOT_FOUND, JobsProblem.OFFLINE)) null else throw error
            }
        if (session() != owner) throw CancellationException("Processing session changed")
        mutable.value = result
        return result
    }
    fun clear() { mutable.value = null }
}

File: ios/Vocal/Processing/ProcessingUsageRepository.swift
import Foundation

struct ProcessingUsage: Decodable, Equatable, Sendable {
  struct Period: Decodable, Equatable, Sendable {
    let key: String
    let start: Date
    let end: Date
    let nextResetAt: Date
  }
  struct Processing: Decodable, Equatable, Sendable {
    let limitSeconds: Double
    let usedSeconds: Double
    let reservedSeconds: Double
    let releasedSeconds: Double
    let remainingSeconds: Double
  }
  struct Availability: Decodable, Equatable, Sendable {
    let status: String
    let reason: String?
  }
  struct Uploads: Decodable, Equatable, Sendable {
    let dailyGrantLimit: Int
    let dailyGrants: Int
    let dailyRemainingGrants: Int
    let dailyResetAt: Date
    let monthlyGrantLimit: Int
    let monthlyGrants: Int
    let monthlyRemainingGrants: Int
    let monthlyByteLimit: Int64
    let confirmedBytes: Int64
    let monthlyRemainingBytes: Int64
    let monthlyResetAt: Date
  }
  struct Storage: Decodable, Equatable, Sendable {
    let limitBytes: Int64
    let retainedBytes: Int64
    let remainingBytes: Int64
  }
  struct Downloads: Decodable, Equatable, Sendable {
    let monthlyGrantLimit: Int
    let monthlyGrants: Int
    let monthlyRemainingGrants: Int
    let monthlyByteLimit: Int64
    let estimatedBytes: Int64
    let monthlyRemainingBytes: Int64
    let monthlyResetAt: Date
  }
  struct EffectiveLimits: Decodable, Equatable, Sendable {
    let maxDurationSeconds: Int
    let maxPreparedAudioBytes: Int64
    let maxClientInputAttempts: Int
    let signedUrlTtlSeconds: Int
  }
  let schemaVersion: Int
  let plan: String
  let policyRevision: Int
  let overrideRevision: Int?
  let effectivePolicySource: String
  let overrideExpiresAt: Date?
  let period: Period
  let processing: Processing
  let uploads: Uploads
  let storage: Storage
  let effectiveLimits: EffectiveLimits
  let downloads: Downloads
  let usageRevision: Int
  let waitingJobs: Int
  let maxWaitingJobs: Int
  let processingJobs: Int
  let maxProcessingJobs: Int
  let availability: Availability
  let checkedAt: Date

  func validate() throws {
    let amounts = [
      processing.limitSeconds, processing.usedSeconds, processing.reservedSeconds,
      processing.releasedSeconds, processing.remainingSeconds,
    ]
    guard
      schemaVersion == 2, plan == "standard", policyRevision >= 0,
      overrideRevision.map({ $0 > 0 }) ?? true, usageRevision >= 0,
      ["global", "account_override"].contains(effectivePolicySource),
      amounts.allSatisfy({ $0.isFinite && $0 >= 0 }),
      processing.remainingSeconds <= processing.limitSeconds,
      uploads.dailyGrantLimit > 0, uploads.monthlyGrantLimit > 0,
      uploads.dailyGrants >= 0, uploads.dailyRemainingGrants >= 0,
      uploads.monthlyGrants >= 0, uploads.monthlyRemainingGrants >= 0,
      uploads.dailyRemainingGrants <= uploads.dailyGrantLimit,
      uploads.monthlyRemainingGrants <= uploads.monthlyGrantLimit,
      uploads.monthlyByteLimit > 0, uploads.confirmedBytes >= 0,
      uploads.monthlyRemainingBytes >= 0,
      uploads.monthlyRemainingBytes <= uploads.monthlyByteLimit,
      storage.limitBytes > 0, storage.retainedBytes >= 0,
      storage.remainingBytes >= 0, storage.remainingBytes <= storage.limitBytes,
      downloads.monthlyGrantLimit > 0, downloads.monthlyGrants >= 0,
      downloads.monthlyRemainingGrants >= 0,
      downloads.monthlyRemainingGrants <= downloads.monthlyGrantLimit,
      downloads.monthlyByteLimit > 0, downloads.estimatedBytes >= 0,
      downloads.monthlyRemainingBytes >= 0,
      downloads.monthlyRemainingBytes <= downloads.monthlyByteLimit,
      effectiveLimits.maxDurationSeconds > 0, effectiveLimits.maxPreparedAudioBytes > 0,
      effectiveLimits.maxClientInputAttempts > 0,
      (1...600).contains(effectiveLimits.signedUrlTtlSeconds),
      waitingJobs >= 0, maxWaitingJobs > 0,
      processingJobs >= 0, maxProcessingJobs > 0, period.start < period.end,
      period.nextResetAt == period.end,
      uploads.monthlyResetAt == period.end, downloads.monthlyResetAt == period.end,
      ["available", "blocked"].contains(availability.status),
      [nil, "paused", "monthly_limit_reached", "waiting_job_limit", "storage_limit_reached"]
        .contains(availability.reason),
      (availability.status == "available") == (availability.reason == nil)
    else { throw JobsFailure.malformedResponse }
  }
}

@MainActor final class ProcessingUsageRepository: ObservableObject {
  @Published private(set) var usage: ProcessingUsage?
  private let api: JobsAPI
  private var owner: String?
  private var receivedAt: Date?
  init(api: JobsAPI) { self.api = api }
  func bind(_ owner: String?) {
    self.owner = owner
    usage = nil
    receivedAt = nil
  }
  func refresh() async {
    guard let captured = owner else { return }
    do {
      let value = try await api.processingUsage()
      try value.validate()
      guard owner == captured else { return }
      usage = value
      receivedAt = Date()
    } catch { if owner == captured { usage = nil } }
  }
  func checkAvailability() throws {
    guard let usage, let receivedAt, abs(receivedAt.timeIntervalSinceNow) < 120 else { return }
    if usage.waitingJobs >= usage.maxWaitingJobs {
      throw JobsFailure.conflict(code: "PROCESSING_LIMIT_REACHED")
    }
    if usage.processing.remainingSeconds <= 0 {
      throw JobsFailure.conflict(code: "PROCESSING_ALLOWANCE_EXHAUSTED")
    }
    switch usage.availability.reason {
    case "monthly_limit_reached":
      throw JobsFailure.conflict(code: "PROCESSING_ALLOWANCE_EXHAUSTED")
    case "waiting_job_limit":
      throw JobsFailure.conflict(code: "PROCESSING_LIMIT_REACHED")
    case "paused":
      throw JobsFailure.conflict(code: "PROCESSING_CAPACITY_UNAVAILABLE")
    default:
      break
    }
  }
}

enum ProcessingMediaMessage {
  static let serverCodes: Set<String> = [
    "MEDIA_TOO_LONG", "MEDIA_TOO_LARGE", "MEDIA_NO_AUDIO",
    "MEDIA_DEFAULT_TRACK_UNAVAILABLE", "MEDIA_UNSUPPORTED", "MEDIA_DURATION_UNKNOWN",
    "YOUTUBE_PLAYLIST_UNSUPPORTED", "YOUTUBE_LIVE_UNSUPPORTED", "PROCESSING_ALLOWANCE_EXHAUSTED",
    "PROCESSING_QUEUE_FULL", "PROCESSING_POLICY_INCOMPATIBLE", "PROCESSING_CAPACITY_UNAVAILABLE",
    "PROCESSING_LIMIT_REACHED",
    "UPLOAD_GRANT_LIMIT_REACHED", "UPLOAD_BYTE_LIMIT_REACHED", "UPLOAD_ATTEMPT_LIMIT_REACHED",
    "RETAINED_STORAGE_LIMIT_REACHED", "DOWNLOAD_RESERVATION_EXPIRED",
    "DOWNLOAD_GRANT_LIMIT_REACHED", "DOWNLOAD_BYTE_LIMIT_REACHED",
    "SERVICE_BANDWIDTH_LIMIT_REACHED",
  ]
  static func key(_ error: Error) -> String? {
    if case JobsFailure.conflict(let code) = error, let code, serverCodes.contains(code) {
      return "media_" + code.lowercased()
    }
    guard let error = error as? AudioInputPreparationError else { return nil }
    switch error {
    case .noAudio: return "media_media_no_audio"
    case .defaultTrackUnavailable: return "media_media_default_track_unavailable"
    case .durationUnknown: return "media_media_duration_unknown"
    case .tooLong: return "media_media_too_long"
    case .invalidSize: return "media_media_too_large"
    case .unsupportedFormat: return "media_media_unsupported"
    case .policyUnavailable: return "media_processing_policy_incompatible"
    case .interrupted: return "media_interrupted"
    default: return nil
    }
  }
}
